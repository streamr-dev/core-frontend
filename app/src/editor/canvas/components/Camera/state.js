import isEditableElement from '$editor/shared/utils/isEditableElement'
import styles from './Camera.pcss'

function clamp(value, min, max) {
    return Math.max(min, Math.min(max, value))
}

function getParentMatching(el, selector) {
    if (!el) { return false }
    if (el.matches(selector)) { return el }
    return getParentMatching(el.parentElement, selector)
}

function scaleToFit({ bounds, fit, maxScale = 1, padding = 20 } = {}) {
    const totalPadding = 2 * padding
    const maxWidth = fit.width - totalPadding
    const maxHeight = fit.height - totalPadding
    return Math.min(maxScale, Math.min(maxWidth / bounds.width, maxHeight / bounds.height))
}

function centerBoundsTo({ bounds, fit, scale }) {
    // vertically & horizontally center content
    const offsetY = fit.height / 2
    const offsetX = fit.width / 2

    return {
        x: -((bounds.x + (bounds.width / 2)) * scale) + offsetX,
        y: -((bounds.y + (bounds.height / 2)) * scale) + offsetY,
    }
}

function toPrecision(v, precision = 0) {
    if (!precision) { return v }
    const p = 10 ** precision
    return Math.round((v * p)) / p
}

const defaultFit = {
    x: 0,
    y: 0,
    scale: 1,
}

function fitBoundsTo({ bounds, fit, padding }) {
    const scale = scaleToFit({
        bounds,
        fit,
        padding,
    })

    if (!scale) { return defaultFit }

    const { x, y } = centerBoundsTo({
        bounds,
        fit,
        scale,
    })

    return {
        x,
        y,
        scale,
    }
}

export function updateScaleState(s, { x, y, scale: newScale }) {
    const { scale: currentScale } = s
    if (!currentScale) { throw new Error(`invalid current scale: ${currentScale}`) }
    if (!newScale) { throw new Error(`invalid new scale: ${newScale}`) }

    // adjust for offset created by scale change
    const ratio = (1 - (newScale / currentScale))
    const x2 = s.x + ((x - s.x) * ratio)
    const y2 = s.y + ((y - s.y) * ratio)

    return {
        ...s,
        scale: newScale,
        x: x2,
        y: y2,
    }
}
const defaultCameraOptions = {
    scaleFactor: 0.1,
    scaleLevels: [
        0.1,
        0.25,
        0.5,
        1.0,
        1.25,
        1.50,
        2.0,
        3.0,
    ],
}

export function createCamera(opts = {}) {
    const { scaleFactor, scaleLevels } = Object.assign({}, defaultCameraOptions, opts)
    const minScale = scaleLevels[0]
    const maxScale = scaleLevels[scaleLevels.length - 1]
    return {
        scale: 1,
        x: 0,
        y: 0,
        elRef: opts.elRef || { current: undefined },
        scaleFactor,
        scaleLevels,
        minScale,
        maxScale,
    }
}

// clamps scale
export function setState(state, v) {
    const nextState = (typeof v === 'function') ? v(state) : v
    const { minScale, maxScale } = nextState
    return {
        ...nextState,
        scale: clamp(nextState.scale, minScale, maxScale),
    }
}

// updates scale of camera, centers scaling on x,y
export function updateScale(state, { x, y, delta }) {
    const { minScale, maxScale, scaleFactor } = state
    const factor = Math.abs(scaleFactor * (delta / 120))
    const actualScaleFactor = delta < 0 ? 1 + factor : 1 - factor
    return setState(state, (s) => (
        updateScaleState(s, {
            x,
            y,
            scale: clamp(s.scale * actualScaleFactor, minScale, maxScale),
        })
    ))
}

// changes current camera position
export function setPosition(state, { x, y }) {
    return setState(state, {
        ...state,
        x,
        y,
    })
}

function getBounds(state) {
    return state.elRef.current.getBoundingClientRect()
}

// set scale to value, centered on middle of camera el
export function setScale(state, value) {
    return setState(state, (s) => {
        const { width, height } = getBounds(s)
        return updateScaleState(s, {
            x: width / 2,
            y: height / 2,
            scale: toPrecision(value, 2),
        })
    })
}

// zoom in to next largest scale level
export function zoomIn(state) {
    const { scale, scaleLevels } = state
    const currentScale = toPrecision(scale, 2)
    const nextLevel = scaleLevels.find((v) => v > currentScale) || scale
    return setScale(state, nextLevel)
}

// zoom out to next largest scale level
export function zoomOut(state) {
    const { scale, scaleLevels } = state
    const currentScale = toPrecision(scale, 2)
    const nextLevel = scaleLevels.slice().reverse().find((v) => v < currentScale) || scale
    return setScale(state, nextLevel)
}

// changes current camera offset i.e. pans camera
export function pan(state, { x = 0, y = 0 } = {}) {
    return setState(state, (s) => ({
        ...s,
        x: s.x + x,
        y: s.y + y,
    }))
}

// fit camera to supplied bounds
export function fitBounds(state, opts) {
    return setState(state, (s) => ({
        ...s,
        ...fitBoundsTo(opts),
    }))
}

// fit supplied bounds to camera element bounds
export function fitView(state, opts) {
    const { width, height } = getBounds(state)
    return fitBounds(state, {
        ...opts,
        fit: {
            width,
            height,
        },
    })
}

// move bounds into center
export function centerBounds(state, { bounds }) {
    const { width, height } = getBounds(state)
    return setState(state, (s) => ({
        ...s,
        ...centerBoundsTo({
            bounds,
            fit: {
                width,
                height,
            },
            scale: s.scale,
        }),
    }))
}

export function areBoundsInView(state, { bounds, padding = 0 }) {
    const { x, y, scale } = state
    const { width, height } = getBounds(state)

    return (
        ((bounds.x * scale) + x) >= padding
        && ((bounds.y * scale) + y) >= padding
        && (((bounds.x + bounds.width) * scale) + x) < (width - padding)
        && (((bounds.y + bounds.height) * scale) + y) < (height - padding)
    )
}

// minimally move bounds into view, if not already in view
export function panIntoViewIfNeeded(state, { bounds, padding }) {
    const inView = areBoundsInView(state, {
        bounds,
        padding,
    })
    if (inView) { return }

    const { x, y, scale } = state
    const { width, height } = getBounds(state)

    const [overLeft, overTop, overRight, overBottom] = [
        padding - ((bounds.x * scale) + x),
        padding - ((bounds.y * scale) + y),
        (((bounds.x + bounds.width) * scale) + x) - (width - padding),
        (((bounds.y + bounds.height) * scale) + y) - (height - padding),
    ].map((v) => Math.max(0, v))

    // if too big to fit on screen rescale and center
    if ((overLeft > 0 && overRight > 0) || (overTop > 0 && overBottom > 0)) {
        return fitView(state, {
            bounds,
            padding,
        })
    }

    // prioritise top & left offsets over right & bottom
    const offsetX = overLeft > 0 ? overLeft : -overRight
    const offsetY = overTop > 0 ? overTop : -overBottom

    return setState(state, (s) => ({
        ...s,
        x: s.x + offsetX,
        y: s.y + offsetY,
    }))
}

// convert bounds in screen/camera coordinates to world coordinates
export function cameraToWorldBounds(state, bounds) {
    const { x, y, scale } = state
    const canvasX = (bounds.x - x) / scale
    const canvasY = (bounds.y - y) / scale
    return {
        x: canvasX,
        y: canvasY,
        width: bounds.width / scale,
        height: bounds.height / scale,
    }
}

// convert point in screen/camera coordinates to world coordinates
export function cameraToWorldPoint(state, point) {
    const { x, y } = cameraToWorldBounds(state, {
        ...point,
        width: 1,
        height: 1,
    })
    return {
        x,
        y,
    }
}

export function eventToWorldPoint(state, event) {
    const point = {
        x: 0,
        y: 0,
    }

    const { elRef } = state
    if (event.target === elRef.current) {
        // if target is camera el then we can just use event offsets
        point.x = event.offsetX
        point.y = event.offsetY
    } else {
        const { left, top } = getBounds(state)
        // otherwise need to calculate offset
        point.x = event.clientX - left
        point.y = event.clientY - top
    }

    const { x, y } = cameraToWorldBounds(state, {
        ...point,
        width: 1,
        height: 1,
    })
    return {
        x,
        y,
    }
}

export function shouldIgnoreEvent(state, event) {
    const { current: el } = state.elRef

    // do not ignore if has cameraControl class
    if (event.target.classList.contains(styles.cameraControl)) { return false }
    // always ignore if has noCameraControl class
    if (event.target.classList.contains(styles.noCameraControl)) { return true }
    // always ignore if target is editable
    if (isEditableElement(event.target)) { return true }

    // do not ignore if is within parent with cameraControl class
    const cameraControlEl = getParentMatching(event.target, `.${styles.cameraControl}`)
    // ignore if is within parent with noCameraControl class
    const noCameraControlEl = getParentMatching(event.target, `.${styles.noCameraControl}`)
    if (cameraControlEl && noCameraControlEl) {
        return cameraControlEl.contains(noCameraControlEl)
    }
    if (cameraControlEl) { return false }
    if (noCameraControlEl) { return false }

    return (
        // ignore if target not within camera el and isn't body
        !el.contains(event.target) && event.target !== document.body
    )
}
