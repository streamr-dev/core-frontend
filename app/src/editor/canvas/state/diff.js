import isEqual from 'lodash/isEqual'
import differenceWith from 'lodash/differenceWith'
import orderBy from 'lodash/orderBy'

const EMPTY = []

function getIsEqualIgnoring(ignoreKeys = new Set()) {
    return (objA, objB) => {
        if (objA === objB) { return true }
        const keysA = Object.keys(objA)
        /* eslint-disable no-continue */
        for (let i = 0; i < keysA.length; i += 1) {
            const key = keysA[i]
            if (ignoreKeys.has(key)) {
                // ignore module internals e.g. subcanvas module
                continue
            }
            if (objA[key] === objB[key]) {
                continue
            }
            const areEqual = isEqual(objA[key], objB[key])
            if (!areEqual) { return false }
        }
        /* eslint-enable no-continue */
        return true
    }
}

const MODULE_IGNORE_KEYS = new Set([
    'modules',
])

const CANVAS_IGNORE_KEYS = new Set([
    'created',
    'updated',
    'modules',
])

const isEqualModule = getIsEqualIgnoring(MODULE_IGNORE_KEYS)
const isEqualCanvasCheck = getIsEqualIgnoring(CANVAS_IGNORE_KEYS)

export function changedModules(canvasA, canvasB) {
    if (!canvasA || !canvasB) { return EMPTY }
    if (canvasA === canvasB || canvasA.modules === canvasB.modules) { return EMPTY }
    const modulesA = orderBy(canvasA.modules || [], 'hash')
    const modulesB = orderBy(canvasB.modules || [], 'hash')
    const AtoB = new Set(differenceWith(modulesA, modulesB, isEqualModule).map(({ hash }) => hash))
    const BtoA = new Set(differenceWith(
        // ignore items already flagged as changed
        modulesB.filter(({ hash }) => !AtoB.has(hash)),
        modulesA.filter(({ hash }) => !AtoB.has(hash)),
        isEqualModule,
    ).map(({ hash }) => hash))
    const result = Array.from(new Set([...AtoB, ...BtoA]))
    if (!result.length) { return EMPTY }
    return result
}

export function isEqualCanvas(canvasA, canvasB) {
    if (canvasA === canvasB) { return true }
    if (!canvasA || !canvasB) { return false }
    const hasChangedModules = changedModules(canvasA, canvasB)
    if (hasChangedModules.length) { return false }
    // don't re-check modules, ignore updated time
    return isEqualCanvasCheck(canvasA, canvasB)
}

