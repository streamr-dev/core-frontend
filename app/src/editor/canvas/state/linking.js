/**
 * Store links between objects (e.g. canvases) in local storage.
 *
 */

const LINK_PREFIX = 'Link'
function validateId(id) {
    if ((typeof id !== 'string' && typeof id !== 'number') || id === '') {
        throw new TypeError(`id must be non-empty string|number. Got ${typeof id}: ${id}`)
    }
}

export function getLink(parentId) {
    validateId(parentId)
    return window.localStorage.getItem(`${LINK_PREFIX}:${parentId}`)
}

export function link(parentId, childId) {
    validateId(parentId)
    validateId(childId)
    return window.localStorage.setItem(`${LINK_PREFIX}:${parentId}`, childId)
}

export function unlink(parentId) {
    validateId(parentId)
    window.localStorage.removeItem(`${LINK_PREFIX}:${parentId}`)
}
