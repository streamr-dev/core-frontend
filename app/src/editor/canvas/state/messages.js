import startCase from 'lodash/startCase'
import sortBy from 'lodash/sortBy'

import routes from '$routes'

import * as CanvasState from './index'

export const LEVELS = [
    'none',
    'info',
    'warn',
    'error',
]

function NoFieldsConfiguredMessage(canvas, moduleHash) {
    const streamModule = CanvasState.getModule(canvas, moduleHash)
    const streamParam = streamModule.params.find(({ name }) => name === 'stream')
    const streamId = streamParam && streamParam.value
    const streamLink = routes.streams.show({ id: streamId })
    return {
        level: 'warn',
        translate: true,
        title: 'editor.canvas.msg.NoFieldsConfiguredTitle',
        content: 'editor.canvas.msg.NoFieldsConfiguredMessage',
        canvasId: canvas.id,
        moduleHash,
        details: {
            streamLink,
        },
    }
}

function NoStreamSelectedMessage(canvas, moduleHash) {
    return {
        level: 'warn',
        title: 'No stream selected',
        content: 'Please select a stream',
        canvasId: canvas.id,
        moduleHash,
    }
}

const EMPTY = []
const STREAM_MODULE_ID = 147
const SEND_TO_STREAM_MODULE_ID = 197

/**
 * Adds messages to Stream & Send to Stream modules
 */

function StreamModuleMesssages(canvas, moduleHash) {
    const m = CanvasState.getModuleIfExists(canvas, moduleHash)
    if (!m) { return EMPTY }

    if (m.id === STREAM_MODULE_ID || m.id === SEND_TO_STREAM_MODULE_ID) {
        const streamParam = m.params.find(({ name }) => name === 'stream')
        const streamId = streamParam && streamParam.value
        if (!streamId) {
            // no stream selected
            return [NoStreamSelectedMessage(canvas, moduleHash)]
        }

        // if no fields:
        // stream module *outputs* will be empty
        // send to stream module *inputs* will be empty
        const target = m.id === STREAM_MODULE_ID ? 'outputs' : 'inputs'
        if (!m[target].length) {
            // did not find any fields
            return [NoFieldsConfiguredMessage(canvas, moduleHash)]
        }

        return EMPTY
    }

    return EMPTY
}

/*
 * Looks for errors in canvas.moduleErrors.
 * e.g. java module compileErrors
 */

function ModuleErrorMesssages(canvas, moduleHash) {
    const m = CanvasState.getModuleIfExists(canvas, moduleHash)
    if (!m) { return EMPTY }
    const { moduleErrors = [] } = canvas
    const currentModuleErrors = moduleErrors.filter(({ module }) => module === moduleHash)
    if (!currentModuleErrors.length) { return EMPTY }
    // sort by ascending line number (if line available)
    return sortBy(currentModuleErrors, 'line').reverse().map(({ type, message, line }) => ({
        level: 'error',
        title: startCase(type),
        content: `${line != null ? `Line ${line}: ` : ''}${message}`,
        canvasId: canvas.id,
        moduleHash,
    }))
}

export function getModuleMessages(canvas, moduleHash) {
    const m = CanvasState.getModuleIfExists(canvas, moduleHash)
    if (!m) { return EMPTY }
    return [
        StreamModuleMesssages(canvas, moduleHash),
        ModuleErrorMesssages(canvas, moduleHash),
    ].flat().map((msg) => ({
        subject: CanvasState.getDisplayName(m),
        ...msg,
    }))
}

export function getMaxLevel(moduleMessages) {
    const levelIndex = moduleMessages.reduce((maxLevel, msg) => (
        Math.max(maxLevel, LEVELS.indexOf(msg.level))
    ), 0)
    return LEVELS[levelIndex]
}

export function getCanvasMessages(canvas) {
    return canvas.modules.map((m) => getModuleMessages(canvas, m.hash)).flat()
}
