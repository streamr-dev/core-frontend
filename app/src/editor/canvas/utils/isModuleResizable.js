// @flow

const RESIZABLE_MODULES = [
    'ChartModule',
    'CommentModule',
    'HeatmapModule',
    'MapModule',
    'LabelModule',
    'TableModule',
    'StreamrTextField',
    'ImageMapModule',
]

export default ({ jsModule, widget }: any) => (
    RESIZABLE_MODULES.includes(jsModule) || RESIZABLE_MODULES.includes(widget)
)
