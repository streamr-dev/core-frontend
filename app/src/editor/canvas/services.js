import axios from 'axios'

import Autosave from '$editor/shared/utils/autosave'
import { emptyCanvas } from './state'

const API = axios.create({
    headers: {
        'Content-Type': 'application/json',
    },
    withCredentials: true,
})

const getData = ({ data }) => data

const canvasesUrl = `${process.env.STREAMR_API_URL}/canvases`
const getModuleURL = `${process.env.STREAMR_URL}/module/jsonGetModule`
const getModuleTreeURL = `${process.env.STREAMR_URL}/module/jsonGetModuleTree`

const AUTOSAVE_DELAY = 3000

async function save(canvas) {
    return API.put(`${canvasesUrl}/${canvas.id}`, canvas).then(getData)
}

export const autosave = Autosave(save, AUTOSAVE_DELAY)

export async function saveNow(canvas, ...args) {
    if (autosave.pending) {
        return autosave.run(canvas, ...args) // do not wait for debounce
    }

    return save(canvas, ...args)
}

async function createCanvas(canvas) {
    return API.post(canvasesUrl, canvas).then(getData)
}

export async function create() {
    return createCanvas(emptyCanvas()) // create new empty
}

export async function moduleHelp({ id }) {
    return API.get(`${process.env.STREAMR_API_URL}/modules/${id}/help`).then(getData)
}

export async function duplicateCanvas(canvas) {
    const savedCanvas = await saveNow(canvas) // ensure canvas saved before duplicating
    return createCanvas(savedCanvas)
}

export async function deleteCanvas({ id }) {
    await autosave.cancel()
    return API.delete(`${canvasesUrl}/${id}`).then(getData)
}

export async function getModuleTree() {
    return API.get(getModuleTreeURL).then(getData)
}

export async function addModule({ id }) {
    const form = new FormData()
    form.append('id', id)
    return API.post(getModuleURL, form).then(getData)
}

export async function loadCanvas({ id }) {
    return API.get(`${canvasesUrl}/${id}`).then(getData)
}

async function startCanvas(canvas, { clearState }) {
    const savedCanvas = await saveNow(canvas)
    return API.post(`${canvasesUrl}/${savedCanvas.id}/start`, {
        clearState: !!clearState,
    }).then(getData)
}

async function startAdhocCanvas(canvas, options = {}) {
    const savedCanvas = await saveNow(canvas)
    const adhocCanvas = await createCanvas({
        ...savedCanvas,
        adhoc: true,
        settings: {
            ...savedCanvas.settings,
            parentCanvasId: canvas.id, // track parent canvas so can return after end
        },
    })
    return startCanvas(adhocCanvas, options)
}

export async function start(canvas, options = {}) {
    const { adhoc } = options
    if (!adhoc) { return startCanvas(canvas, options) }
    return startAdhocCanvas(canvas, options)
}

export async function stop(canvas) {
    return API.post(`${canvasesUrl}/${canvas.id}/stop`).then(getData)
}
