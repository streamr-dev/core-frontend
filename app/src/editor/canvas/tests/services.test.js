import uniqueId from 'lodash/uniqueId'
import { setupAuthorizationHeader, loadModuleDefinition } from '$editor/shared/tests/utils'
import * as sharedServices from '$editor/shared/services'

import * as Services from '../services'
import * as State from '../state'
import { canvasMatcher } from './utils'

describe('Canvas Services', () => {
    let teardown

    beforeAll(async () => {
        teardown = await setupAuthorizationHeader()
    }, 60000) // service can take some time to respond initially

    afterAll(async () => {
        await Services.deleteAllCanvases()
        await teardown()
    })

    describe('create', () => {
        it('can create canvases', async () => {
            const canvases = await Services.loadCanvases()
            expect(canvases.length).toBe(0)
            const canvas = await Services.create()
            expect(canvas).toMatchSnapshot(canvasMatcher)

            expect(canvas).toHaveProperty('id')
            expect(canvas.modules.length).toBe(0)
            const canvasesAfterCreate = await Services.loadCanvases()
            expect(canvasesAfterCreate.length).toBe(1)
            expect(canvasesAfterCreate).toContainEqual({
                ...canvas,
                created: canvasMatcher.created,
                updated: canvasMatcher.updated,
            })
        })
        it('deduplicates canvas name', async () => {
            const canvas = await Services.create()
            const canvas2 = await Services.create()
            expect(canvas.name).not.toEqual(canvas2.name)
        })

        it('can take canvas name', async () => {
            const name1 = `test${uniqueId()}`
            const name2 = `test${uniqueId()}`
            const canvas1 = await Services.create({ name: name1 })
            await Services.create({ name: name1 })
            const canvas3 = await Services.create({ name: name2 })
            expect(canvas1).toMatchObject({
                name: name1,
            })

            expect(canvas3).toMatchObject({
                name: name2,
            })
        })
    })

    describe('loadCanvas', () => {
        it('errors if no canvas', () => {
            expect(Services.loadCanvas({
                id: 'invalid',
            })).rejects.toThrow()
        })

        it('can load a canvas', async () => {
            const canvas = await Services.create()
            const { id } = canvas
            const loadedCanvas = await Services.loadCanvas({ id })
            expect(loadedCanvas).toMatchObject({
                ...canvas,
                // loading canvas gives different created/updated time than create ???
                created: canvasMatcher.created,
                updated: canvasMatcher.updated,
            })
        })
    })

    describe('deleteCanvas', () => {
        it('errors if no canvas', () => {
            expect(Services.deleteCanvas({
                id: 'invalid',
            })).rejects.toThrow()
        })

        it('can delete a canvas', async () => {
            const canvas = await Services.create()
            const { id } = canvas
            await Services.deleteCanvas({ id })
            await expect(Services.loadCanvas({ id })).rejects.toThrow()
        })
    })

    describe('duplicateCanvas', () => {
        it('errors if no canvas', () => {
            expect(Services.duplicateCanvas({
                id: 'invalid',
            })).rejects.toThrow()
        })

        it('can duplicate a canvas', async () => {
            const canvas = await Services.create()
            const duplicateCanvas = await Services.duplicateCanvas(canvas)
            expect(duplicateCanvas.id).not.toEqual(canvas.id)
            expect(duplicateCanvas).toMatchObject({
                ...canvas,
                ...canvasMatcher,
            })
            expect(duplicateCanvas.name.startsWith(canvas.name))
        })

        it('deduplicates canvas name', async () => {
            const name = `test${uniqueId()}`
            const canvas = await Services.create({
                name,
            })
            const duplicateCanvas = await Services.duplicateCanvas(canvas)
            expect(duplicateCanvas).toMatchObject({
                ...canvas,
                ...canvasMatcher,
                name: `${name} Copy`,
            })
            const duplicateCanvas2 = await Services.duplicateCanvas(duplicateCanvas)
            expect(duplicateCanvas2).toMatchObject({
                ...canvas,
                ...canvasMatcher,
                name: `${name} Copy 02`,
            })
            // test works with double-digit numbers
            const duplicateCanvas3 = await Services.create({
                ...canvas,
                name: `${name} Copy 10`,
            })
            expect(duplicateCanvas3).toMatchObject({
                ...canvas,
                ...canvasMatcher,
                name: `${name} Copy 10`, // should not change from what was passed in
            })
            const duplicateCanvas4 = await Services.duplicateCanvas(duplicateCanvas3)
            expect(duplicateCanvas4).toMatchObject({
                ...canvas,
                ...canvasMatcher,
                name: `${name} Copy 11`,
            })
        })

        it('can duplicate a running canvas', async () => {
            const canvas = await Services.create()
            const startedCanvas = await Services.start(canvas)

            const duplicateCanvas = await Services.duplicateCanvas(startedCanvas)
            expect(duplicateCanvas).toMatchObject({
                ...canvas,
                ...canvasMatcher,
                state: State.RunStates.Stopped,
            })
            await Services.stop(startedCanvas)
        })
    })

    describe('start/stop canvas', () => {
        it('errors if no canvas', () => {
            expect(Services.start({
                id: 'invalid',
            })).rejects.toThrow()
            expect(Services.stop({
                id: 'invalid',
            })).rejects.toThrow()
        })

        it('can start & stop a canvas', async () => {
            const canvas = await Services.create()
            expect(State.isRunning(canvas)).toBeFalsy()

            const startedCanvas = await Services.start(canvas)
            expect(State.isRunning(startedCanvas)).toBeTruthy()
            expect(startedCanvas).toMatchObject({
                ...canvas,
                ...canvasMatcher,
                id: canvas.id,
                state: State.RunStates.Running,
                startedById: expect.any(Number),
            })

            // can't start a running canvas
            await expect(Services.start(canvas)).rejects.toThrow()

            const stoppedCanvas = await Services.stop(canvas)
            expect(State.isRunning(stoppedCanvas)).toBeFalsy()
            expect(stoppedCanvas).toMatchObject({
                ...startedCanvas,
                ...canvasMatcher,
                id: canvas.id,
                state: State.RunStates.Stopped,
                startedById: expect.any(Number),
            })

            // can't stop a stopped canvas
            await expect(Services.stop(canvas)).rejects.toThrow()
        })
    })

    describe('Adding Modules', () => {
        it('creates modules with compatible hash values', async () => {
            // add a clock
            let canvas = State.emptyCanvas()
            canvas = State.addModule(canvas, await loadModuleDefinition('Clock'))
            const clock = canvas.modules.find((m) => m.name === 'Clock')
            expect(clock).toBeTruthy()

            const savedCanvas = await Services.create(canvas)
            const savedClock = savedCanvas.modules.find((m) => m.name === 'Clock')
            expect(savedClock).toBeTruthy()
            // ensure hash value wasn't messed up by type conversion
            expect(savedClock.hash).toEqual(clock.hash)
        })

        it('creates modules with unique hash', async () => {
            let canvas = State.emptyCanvas()
            const ClockDefn = await loadModuleDefinition('Clock')
            // add a few clocks in quick succession
            canvas = State.addModule(canvas, ClockDefn)
            canvas = State.addModule(canvas, ClockDefn)
            canvas = State.addModule(canvas, ClockDefn)
            const hashes = new Set(canvas.modules.map(({ hash }) => hash))
            // make sure they all have unique hashes
            expect(hashes.size).toEqual(3)
        })
    })

    describe('Replacing Module Definition', () => {
        it('maintains connections after loading new definition for http request', async () => {
            // connect something to http request's body input
            let canvas = State.emptyCanvas()
            canvas = State.addModule(canvas, await loadModuleDefinition('HTTP Request'))
            canvas = State.addModule(canvas, await loadModuleDefinition('ConstantText'))
            const httpRequest = canvas.modules.find((m) => m.name === 'HTTP Request')
            const text = canvas.modules.find((m) => m.name === 'ConstantText')
            const httpRequestBodyIn = State.findModulePort(canvas, httpRequest.hash, (p) => p.name === 'body')
            const textOut = State.findModulePort(canvas, text.hash, (p) => p.name === 'out')

            const savedCanvas = State.updateCanvas(await Services.create(canvas))
            canvas = State.updateCanvas(State.connectPorts(canvas, textOut.id, httpRequestBodyIn.id))
            expect(State.arePortsConnected(canvas, textOut.id, httpRequestBodyIn.id)).toBeTruthy()
            const httpRequestVerb = State.findModulePort(savedCanvas, httpRequest.hash, (p) => p.name === 'verb')

            // change verb to something else
            canvas = State.updateCanvas(State.setPortUserValue(canvas, httpRequestVerb.id, 'PUT'))
            const updatedHttpRequest = canvas.modules.find((m) => m.name === 'HTTP Request')
            const updatedHttpRequestFromServer = await sharedServices.getModule({
                id: updatedHttpRequest.id,
                configuration: updatedHttpRequest,
            })

            // replace with new definition after changing verb
            canvas = State.updateCanvas(State.replaceModule(canvas, updatedHttpRequestFromServer))

            // ensure connections maintained
            const newHttpRequestBodyIn = State.findModulePort(canvas, httpRequest.hash, (p) => p.name === 'body')
            const newTextOut = State.findModulePort(canvas, text.hash, (p) => p.name === 'out')
            expect(State.arePortsConnected(canvas, newTextOut.id, newHttpRequestBodyIn.id)).toBeTruthy()
        })

        it('maintains connections after loading new definition for canvas module', async () => {
            let subcanvas1 = State.emptyCanvas()
            subcanvas1 = State.addModule(subcanvas1, await loadModuleDefinition('ConstantText'))
            const text1 = subcanvas1.modules.find((m) => m.name === 'ConstantText')
            const text1Out = State.findModulePort(subcanvas1, text1.hash, (p) => p.name === 'out')
            const text1In = State.findModulePort(subcanvas1, text1.hash, (p) => p.name === 'str')

            subcanvas1 = State.setPortOptions(subcanvas1, text1Out.id, {
                export: true,
            })

            subcanvas1 = State.setPortOptions(subcanvas1, text1In.id, {
                export: true,
            })

            subcanvas1 = State.updateCanvas(await Services.create(subcanvas1))

            const subcanvas2 = State.updateCanvas(await Services.duplicateCanvas(subcanvas1))

            // connect something to http request's body input
            let canvas = State.emptyCanvas()
            canvas = State.addModule(canvas, await loadModuleDefinition('Table'))
            canvas = State.addModule(canvas, await loadModuleDefinition('Canvas'))

            const table = canvas.modules.find((m) => m.name === 'Table')
            const canvasModule = canvas.modules.find((m) => m.jsModule === 'CanvasModule')

            // select the subcanvas
            const canvasPort = State.findModulePort(canvas, canvasModule.hash, (p) => p.name === 'canvas')

            canvas = State.updatePort(canvas, canvasPort.id, (port) => ({
                ...port,
                value: subcanvas1.id,
            }))

            canvas = State.updateCanvas(await Services.create(canvas))

            const canvasModuleOut1 = State.findModulePort(canvas, canvasModule.hash, (p) => p.name === 'out')
            const tableIn1 = State.findModulePort(canvas, table.hash, (p) => p.variadic.index === 1)
            // connect canvas module out to table in1 (creates table in2)
            canvas = State.updateCanvas(State.connectPorts(canvas, canvasModuleOut1.id, tableIn1.id))
            const tableIn2 = State.findModulePort(canvas, table.hash, (p) => p.variadic.index === 2)
            // connect canvas module out to table in2
            canvas = State.updateCanvas(State.connectPorts(canvas, canvasModuleOut1.id, tableIn2.id))
            // disconnect canvas module out from table in1, leaves first port empty
            canvas = State.updateCanvas(State.disconnectPorts(canvas, canvasModuleOut1.id, tableIn1.id))

            // change subcanvas
            canvas = State.updateCanvas(State.updatePort(canvas, canvasPort.id, (port) => ({
                ...port,
                value: subcanvas2.id,
            })))

            const updatedCanvasModule = canvas.modules.find((m) => m.jsModule === 'CanvasModule')
            const updatedCanvasModuleFromServer = await sharedServices.getModule({
                id: updatedCanvasModule.id,
                configuration: updatedCanvasModule,
            })

            // replace with new definition
            canvas = State.replaceModule(canvas, updatedCanvasModuleFromServer)
            canvas = State.updateCanvas(canvas)

            // ensure connections maintained
            const newTableIn2 = State.findModulePort(canvas, table.hash, (p) => p.variadic.index === 2)
            const newCanvasModuleOut1 = State.findModulePort(canvas, canvasModule.hash, (p) => p.name === 'out')
            expect(State.arePortsConnected(canvas, newCanvasModuleOut1.id, newTableIn2.id)).toBeTruthy()
        })
    })
})
