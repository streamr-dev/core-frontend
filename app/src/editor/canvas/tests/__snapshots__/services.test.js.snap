// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas Services create can create canvases 1`] = `
Object {
  "adhoc": false,
  "created": Any<String>,
  "hasExports": false,
  "id": Any<String>,
  "modules": Array [],
  "name": Any<String>,
  "serialized": false,
  "settings": Object {
    "editorState": Object {
      "runTab": "#tab-realtime",
    },
  },
  "startedById": null,
  "state": "STOPPED",
  "uiChannel": ObjectContaining {
    "id": Any<String>,
  },
  "updated": Any<String>,
}
`;
